File: Solutions/0673.Number-of-Longest-Increasing-Subsequence.java
Given an unsorted array of integers, find the number of longest increasing subsequence.

Example 1:
Input: [1,3,5,4,7]
Output: 2
Explanation: The two longest increasing subsequence are [1, 3, 4, 7] and [1, 3, 5, 7].



class Solution {
    public int findNumberOfLIS(int[] nums) {
        // dp[i] = the number of longest increasing subseq until the ith number
        // if nums[j] > nums[i] for j>i: then dp[j] += dp[i]
        // if nums[j] < nums[i] for j > i: then dp[j] = dp[j - 1]
        
        int lens = nums.length;
        if (lens <= 1) return lens;
        
        int[] dp = new int[lens];   // dp[i]=the maxLen ended with i
        int[] cnt = new int[lens];  // cnt[i]= the cnt of maxLen ended with i
        
        for (int i = 0; i < lens; i++) {
            dp[i] = 1;
            cnt[i] = 1;
        }
        
        for (int j = 1; j < lens; j++) {
            for (int i = 0; i < j; i++) {
                if (nums[j] > nums[i]) {
                    if (dp[j] <= dp[i]) {  // first time to renew dp[j]
                        dp[j] = dp[i] + 1;
                        cnt[j] = cnt[i];
                    } else if (dp[j] == dp[i] + 1) {
                        dp[j] = dp[i] + 1;
                        cnt[j] += cnt[i];
                    }
                } 
            }
        }
        
        System.out.println(Arrays.toString(dp));
        System.out.println(Arrays.toString(cnt));
        int maxLen = 0;
        int cntofMaxLen = 0;
        for (int i = 0; i < lens; i++) {
            maxLen = Math.max(maxLen, dp[i]);
        }
        
        for (int i = 0; i < lens; i++) {
            if (dp[i] == maxLen) {
                cntofMaxLen += cnt[i];
            }
        }
        
        return cntofMaxLen;
    }
}

File: Solutions/0673.Number-of-Longest-Increasing-Subsequence.py
Given an unsorted array of integers, find the number of longest increasing subsequence.

Example 1:
Input: [1,3,5,4,7]
Output: 2
Explanation: The two longest increasing subsequence are [1, 3, 4, 7] and [1, 3, 5, 7].


class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        lens = len(nums)
        dp = [1] * lens     # 以i为结尾的最大的长度
        cnt = [1] * lens    # 以i为结尾的最大的长度的个数
        
        for j in range(1, lens):
            for i in range(j):
                if nums[j] > nums[i]:   # 如果包括等于的话，那么[2,2,2]的输出就是错的
                    if dp[j] <= dp[i]:  # meaning this is the first time renewing dp[j], or the first time dp[j]遇到了一个非常大的dp[i]
                        dp[j] = dp[i] + 1
                        cnt[j] = cnt[i]
                        
                    elif dp[j] == dp[i] + 1:    # meaning dp[j]是dp[i]的下一个递增元素
                        cnt[j] += cnt[i]
                        
        print(dp)
        print(cnt)
        maxLen = max(dp)
        res = 0
        for i, length in enumerate(dp):
            if length == maxLen:
                res += cnt[i]
        
        return res
    
    
"""    
Solution 2: segment tree - O(nlogn)   
Suppose we knew for each length L-1, the number of sequences with length L-1 ending in x. 
Then when considering the next element of nums, updating our knowledge hinges on knowing the number of sequences with length L-1 ending in x < y. 
This type of query over an interval is a natural fit for using some sort of tree.
"""
